### ISSUE TITLE
SQL extensions for Nessie catalog

### ISSUE THREAD

**Issue by rymurr** [apache/iceberg#1808]
Currently when using Nessie and Iceberg the only way to perform operations on branches is to use the [CLI ](https://projectnessie.org/tools/cli/) as per the example in the [demo](https://github.com/projectnessie/nessie/tree/main/python/demo). Now that Iceberg supports custom SQL extensions it would be good to add some commands to allow manipulation of branches from Iceberg/Spark directly.

The operations that we propose support for are:

1. Specify the branch/context that operations should take place on. This would change the `reference` in an existing Nessie catalog. Most natural seems to be something like `USE CATALOG REFERENCE <refName> [AT <timestamp`>] where `refName` could be a Branch name, a Tag name or a specific Hash. The optional `timestamp` would allow for time travel on a branch or tag only.

2. create/delete operations on branches. This would not affect the catalog directly but would create or delete a branch. Optionally create can switch to that branch (eg via 1. above). As a strawman I propose `ALTER CATALOG CREATE [BRANCH|TAG] <refName> [AT <hash>|<ref>]` and  `ALTER CATALOG DROP [BRANCH|TAG] <refName>`. There are some issues to work out around deleting the current catalog branch and concurrency control via the `expectedHash` in teh cli

3. merge/assign/cherry-pick operations. These operations are used to move commits between branches. For now I would like to focus on the merge operation. Which would take all commits on Branch A from the HEAD to a common ancestor and move them to Branch B and is roughly analogous to `git merge`. Note: currently Nessie merge fails with no changes if there is a conflict. This could be expressed as `ALTER CATALOG MERGE BRANCH <refName> [<toRefName>]. Here `refName` is the branch we want to merge from and `toRefName` is the branch we want to merge onto. This is optional and the command will use the current branch (as set by 1.) if not provided. The cherry pick operation could be similar but is slightly complicated as one has to specify a contiguous set of hashes. As above some more thinking around concurrency control is needed.

These operations would have to fail gracefully when using a catalog that doesn't support branching.

Hopefully this is enough to start a discussion on. I would appreciate any feedback! If easier I can split the 3 options into 3 subtasks.

**Comment by rdblue** [comment]
I would add `SHOW CATALOG REFERENCE` or some form of `git status` to let users know where they are as well.

**Comment by rdblue** [comment]
I'm not sure that I would use `ALTER CATALOG`, since that seems to imply that the catalog config is being altered, not that the catalog reference is changing. I would probably opt for simpler commands, like `DROP CATALOG BRANCH some_branch` or `CREATE CATALOG TAG new_name`.

I'm also curious about what `MERGE` means. Is that like rebasing another branch's changes onto the current? If not rebase, what would happen?

**Comment by rymurr** [comment]
> I would add `SHOW CATALOG REFERENCE` or some form of `git status` to let users know where they are as well.

Excellent idea. I've also been toying with the idea of showing history as a table. eg `SHOW LOG [ref]` should show a table equivalent of `git log`. I would like to be able to select from it though... eg `SELECT * FROM catalog._meta_.log WHERE commitTime < timestamp and committer = 'rymurr'`. I am not sure how to match these two ideas up.

**Comment by rymurr** [comment]
> I'm also curious about what `MERGE` means. Is that like rebasing another branch's changes onto the current? If not rebase, what would happen?

Thats correct, merge is probably closest to a fast forward rebase except history isn't re-written. If you had
```
 C    E
 |    |
 B    D
  \   /
    A
```
Then merging the right branch into the left would give:
```
 E
 |
 D
 |
 C    E
 |    |
 B    D
  \   /
    A
```
Merge always looks for a common ancestor and transplants commits across. It doesn't create a merge commit with multiple parents like git would. Nessie always keeps a linear history.

If there is no common ancestor or there is a conflict the operation would fail (atomically). Currently the failure report is simply 'merge failed'. There is work on the server in flight to make the failure more clear and actionable. Currently conflict is defined at the table level and for iceberg is simply: 'Table A on branch X has a different current snapshot than on branch Y'. It could be a really simple resolution (eg 2 appends) that could eventually be handled silently or something more complicated: conditional delete and schema change which drops a column from the conditional delete. Which would need more direct human intervention to resolve.

**Comment by shangshengtung** [comment]
Great thought.  Is USE CATALOG BRANCH clearer If only branches are mutable (and USE CATALOG REFERENCE for tag or hash)?

**Comment by shangshengtung** [comment]
Does it need a LIST CATALOG REFERENCE (some form of git branch)?

**Comment by rymurr** [comment]
I think thats a good idea @shangshengtung. Any thoughts about the rest? I am starting to like `CREATE/DROP BRANCH/TAG` more over the `ALTER` syntax

**Comment by shangshengtung** [comment]
How do you think on the time travel syntax of sql statements like (AT|BEFORE, AS OF) to refer a past time?

**Comment by rdblue** [comment]
> How do you think on the time travel syntax of sql statements like (AT|BEFORE, AS OF) to refer a past time?

I really prefer this because the engine is responsible for converting the given timestamp into a specific instant. That's much easier than Iceberg trying to match the same behavior in each engine. It is less important for `AS OF VERSION` because that gets passed directly without modification, but `AS OF TIMESTAMP` support in Spark would be great.

**Comment by rymurr** [comment]
> `AS OF TIMESTAMP` support in Spark would be great.

I agree. One thing to watch out for: if the catalog is a nessie catalog timetravel is via the nessie branching and in Hive/Hadoop catalogs its via snapshot ids. We just have to make sure the same syntax works for both and is not surprising

**Comment by rymurr** [comment]
@rdblue any preference for where this component would live?I have an initial impl done for create tag/branch in with the rest of the sql extensions currently but not sure if its best there. I can create a nessie extensions module if you prefer but there will be some duplication w/ the existing extensions module.

**Comment by rdblue** [comment]
I'm thinking that the SQL extensions should live in Spark, but that we can hopefully define an interface so that they are not specific to Nessie? It would be great to be able to enable these for any catalog that has a similar way of tracking branches. I'm not sure that others necessarily would, but it seems like a good idea not to have Nessie-specific code in the Spark extensions module.

**Comment by rymurr** [comment]
> but it seems like a good idea not to have Nessie-specific code in the Spark extensions module.

Very much agree. Definitely the thing I like least about #2269 

I was thinking about a `BranchingCatalog` or `BranchSupport` or something interface which would add CRUD operations for references to a `Catalog`. The only current implementer would be the `NessieCatalog` but then the spark extensions don't need to know about Nessie and #2269 would be simplified by being able to call Catalog methods directly rather than building a Nessie Client in the extension.

I will open up a PR for that to keep it separate from #2269
